data_structures/sparse_table.cpp

```diff
@@ -0,0 +1,135 @@
+ * `gcd()`, `lcm()`, and `max()` by changing a few lines of code.
+ */
+
+#include <array>     /// for std::array
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+
+/**
+ * @namespace data_structures
+ * @brief Data Structures algorithms
+ */
+namespace data_structures {
+
+/**
+ * @namespace sparse_table
+ * @brief Functions for Implementation of [Sparse
+ * Table](https://brilliant.org/wiki/sparse-table/)
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table for `min()` as their invariant
+ * function, for the given array `A`. The answer to queries are stored in the
+ * array ST.
+ */
+constexpr uint32_t N = 12345;  ///< the maximum size of the array.
+constexpr uint8_t M = 14;      ///< ceil(log2(N)).
+
+struct Sparse_table {
+    size_t n = 0;  ///< size of input array.
+
+    /** @warning check if `N` is not less than `n`. if so, manually increase the
+     * value of N */
+
+    std::array<int64_t, N> A = {};  ///< input array to perform RMQ.
+    std::array<std::array<int64_t, N>, M> ST =
+        {};  ///< the sparse table storing `min()` values for given interval.
+    std::array<int64_t, N> LOG = {};  ///< where floor(log2(i)) are precomputed.
+
+    /**
+     * @brief Builds the sparse table for computing min/max/gcd/lcm/...etc
+     * for any contiguous sub-segment of the array.This is an example of
+     * computing the index of the minimum value.
+     * @return void
+     * @complexity: O(n.log(n))
+     */
+    void buildST() {
+        LOG[0] = -1;
+
+        for (size_t i = 0; i < n; ++i) {
+            ST[0][i] = static_cast<int64_t>(i);
+            LOG[i + 1] = LOG[i] + !(i & (i + 1));
+        }
+
+        for (size_t j = 1; static_cast<size_t>(1 << j) <= n; ++j) {
+            for (size_t i = 0; static_cast<size_t>(i + (1 << j)) <= n; ++i) {
+                int64_t x = ST[j - 1][i];
+                int64_t y = ST[j - 1][i + (1 << (j - 1))];
+
+                ST[j][i] = (A[x] <= A[y] ? x : y);
+            }
+        }
+    }
+
+    /**
+     * @brief Queries the sparse table for the value of the interval [l, r]
+     * (i.e. from l to r inclusive).
+     * @param l the left index of the range (inclusive).
+     * @param r the right index of the range (inclusive).
+     * @return the computed value of the given interval.
+     * @complexity: O(1)
+     */
+    int64_t query(int64_t l, int64_t r) {
+        int64_t g = LOG[r - l + 1];  ///< smallest power of 2 covering [l,r]
+        int64_t x = ST[g][l];
+        int64_t y = ST[g][r - (1 << g) + 1];
+        return (A[x] <= A[y] ? x : y);
+    }
+};
+}  // namespace sparse_table
+}  // namespace data_structures
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    std::array<int64_t, 10> testcase = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+    size_t testcase_size = sizeof(testcase) / sizeof(testcase[0]);
+
+    data_structures::sparse_table::Sparse_table st{};
```

Comment:
Maybe provide a description? The more description you provide, the better understanding for the readers.